src/client/Client.cc

```diff
@@ -11233,8 +11233,14 @@ int64_t Client::_preadv_pwritev_locked(Fh *fh, const struct iovec *iov,
     ceph_assert(ceph_mutex_is_locked_by_me(client_lock));
 
 #if defined(__linux__) && defined(O_PATH)
-    if (fh->flags & O_PATH)
-        return -CEPHFS_EBADF;
+    if (fh->flags & O_PATH) {
+      if(onfinish != nullptr) {
+        client_lock.unlock();
+        onfinish->complete(-CEPHFS_EBADF);
+        client_lock.lock();
+      }
+      return -CEPHFS_EBADF;
```

Comment:
Should be `return 0`.

For async APIs, the actual errno (or 0, in case of success) should be passed to the callback. The API itself should return 0 (unless something really out of the ordinary happened, in which case the _return value` would be non-zero.

Reply:
That's the responsibility of the caller, so, yes.